In [54]:
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.pylab as py

In [ ]:
# 한글 폰트 불러오기

!apt -qq -y install fonts-nanum > /dev/null

import matplotlib.font_manager as fm

fontpath = '/usr/share/fonts/truetype/nanum/NanumBarunGothic.ttf'

font = fm.FontProperties(fname=fontpath, size=9)

fm._rebuild()

# 그래프에 retina display 적용

%config InlineBackend.figure_format = 'retina'

# Colab 의 한글 폰트 설정

plt.rc('font', family='NanumBarunGothic')

In [55]:
patientinfo = pd.read_csv('./drive/My Drive/Colab Notebooks/Warmingup_Project/corona_visualization/PatientInfo.csv')
patientinfo

patient_id  global_num     sex  ... released_date deceased_date     state
0     1000000001         2.0    male  ...    2020-02-05           NaN  released
1     1000000002         5.0    male  ...    2020-03-02           NaN  released
2     1000000003         6.0    male  ...    2020-02-19           NaN  released
3     1000000004         7.0    male  ...    2020-02-15           NaN  released
4     1000000005         9.0  female  ...    2020-02-24           NaN  released
...          ...         ...     ...  ...           ...           ...       ...
3999  7000000010         NaN  female  ...    2020-04-21           NaN  released
4000  7000000011         NaN    male  ...           NaN           NaN  isolated
4001  7000000012         NaN  female  ...    2020-05-05           NaN  released
4002  7000000013         NaN  female  ...    2020-04-26           NaN  released
4003  7000000014         NaN  female  ...           NaN           NaN  isolated

[4004 rows x 18 columns]

# 5. 지역별 신규 확진자가 많았던 날의 감염 원인
* 영남지방 신규 확진자 최대인 날짜 : 2020-02-29 , 최대 신규 확진자 수 : 763 명
* 호남지방 신규 확진자 최대인 날짜 : 2020-04-01 , 최대 신규 확진자 수 : 8 명
* 호서지방 신규 확진자 최대인 날짜 : 2020-02-28 , 최대 신규 확진자 수 : 28 명
* 수도권 신규 확진자 최대인 날짜:  2020-03-11 , 최대 신규 확진자 수 : 76 명
* 관동지방 신규 확진자 최대인 날짜:  2020-03-01 , 최대 신규 확진자 수 : 8 명

In [56]:
# 각 지역별로 데이터 프레임 저장
youngnam_df =  patientinfo[(patientinfo['province'] == 'Busan') | (patientinfo['province'] == 'Daegu') | (patientinfo['province'] == 'Gyeongsangbuk-do') | 
                           (patientinfo['province'] == 'Gyeongsangnam-do') | (patientinfo['province'] == 'Ulsan')]

honam_df = patientinfo[(patientinfo['province'] == 'Gwangju') | (patientinfo['province'] == 'Jeollabuk-do') | (patientinfo['province'] == 'Jeollanam-do')]
hosae_df = patientinfo[(patientinfo['province'] == 'Chungcheongbuk-do') | (patientinfo['province'] == 'Chungcheongnam-do') | (patientinfo['province'] == 'Daejeon') | 
                       (patientinfo['province'] == 'Sejong')]                                                                                                               
sudo_df = patientinfo[(patientinfo['province'] == 'Gyeonggi-do') | (patientinfo['province'] == 'Incheon') | (patientinfo['province'] == 'Seoul')]
gandong_df = patientinfo[patientinfo['province'] == 'Gangwon-do']

In [57]:
youngnam_df

patient_id  global_num     sex  ... released_date deceased_date     state
863   1100000001         NaN    male  ...           NaN           NaN  released
864   1100000002         NaN  female  ...           NaN           NaN  released
865   1100000003         NaN  female  ...           NaN           NaN  isolated
866   1100000004         NaN  female  ...           NaN           NaN  released
867   1100000005         NaN  female  ...           NaN           NaN  released
...          ...         ...     ...  ...           ...           ...       ...
3985  6100000118         NaN    male  ...           NaN           NaN  isolated
3986  6100000119         NaN  female  ...           NaN           NaN  released
3987  6100000120         NaN  female  ...           NaN           NaN  released
3988  6100000121         NaN    male  ...           NaN           NaN  isolated
3989  6100000122         NaN    male  ...           NaN           NaN  released

[1626 rows x 18 columns]

In [ ]:
# 각 지역별 데이터 프레임을 신규 확진자가 많이 발생한 날짜의 최대 감염 원인 출력

In [58]:
# 영남 지방 감염 원인

# etc                                      319
# contact with patient                     219
# overseas inflow                           91
# Shincheonji Church                        78
# Onchun Church                             33
# Bonghwa Pureun Nursing Home               31
# Cheongdo Daenam Hospital                  21
# Gyeongsan Seorin Nursing Home             15
# Gyeongsan Jeil Silver Town                12
# Milal Shelter                             11
# Gyeongsan Cham Joeun Community Center     10
# Geochang Church                            6
# Changnyeong Coin Karaoke                   4
# Suyeong-gu Kindergarten                    3
# Itaewon Clubs                              2
# Pilgrimage to Israel                       2

# 영남지방 2020-02-29 최대 감염원인
print('영남 확진자 최대일: ',youngnam_df['confirmed_date'].value_counts().idxmax())   # 영남 확진자 최대일:  2020-03-05
print('영남 최대 확진자 :', youngnam_df['confirmed_date'].value_counts().max(),'명')  # 영남 최대 확진자 : 124 명
# 

영남 확진자 최대일:  2020-03-05
영남 최대 확진자 : 124 명


In [59]:
youngnam_df[youngnam_df['confirmed_date']=='2020-03-05']['infection_case'].value_counts()

etc                                      10
contact with patient                      9
Bonghwa Pureun Nursing Home               7
Shincheonji Church                        4
Gyeongsan Cham Joeun Community Center     3
Changnyeong Coin Karaoke                  1
Gyeongsan Jeil Silver Town                1
Name: infection_case, dtype: int64

In [63]:
print(honam_df['infection_case'].value_counts())
# hosae_cause = hosae_df
# sudo_caues = sudo_df
# gandong_cause = gandong_df

overseas inflow         35
contact with patient    18
etc                     14
Shincheonji Church       2
Itaewon Clubs            1
Name: infection_case, dtype: int64


In [ ]:
# 호남 지방 감염 원인

# overseas inflow         35
# contact with patient    18
# etc                     14
# Shincheonji Church       2
# Itaewon Clubs            1

# 호남지방 2020-04-01 최대 김염원인
# honam_df[honam_df['confirmed_date'] == '2020-04-01']['infection_case'].value_counts()
honam_df.groupby('confirmed_date').size()

In [61]:
# 호서 지방 감염 원인
# hosae_df['infection_case'].value_counts()
# contact with patient                131
# etc                                  57
# overseas inflow                      40
# gym facility in Cheonan              28
# Ministry of Oceans and Fisheries     27
# gym facility in Sejong                4
# Shincheonji Church                    2
# Itaewon Clubs                         1
# River of Grace Community Church       1




In [62]:
# 수도권 최대 감염 원인
sudo_df['infection_case'].value_counts()

contact with patient                793
overseas inflow                     514
etc                                 199
Itaewon Clubs                       150
Guro-gu Call Center                 112
Coupang Logistics Center             70
Dongan Church                        17
Eunpyeong St. Mary's Hospital        16
Shincheonji Church                   15
Seongdong-gu APT                     13
KB Life Insurance                     8
Samsung Medical Center                7
Guri Collective Infection             6
Yeongdeungpo Learning Institute       2
Ministry of Oceans and Fisheries      1
gym facility in Cheonan               1
Name: infection_case, dtype: int64

In [30]:
# 관동 지방 최대 감염 원인
gandong_df['infection_case'].value_counts()

contact with patient       20
overseas inflow            15
Shincheonji Church         10
etc                         8
gym facility in Cheonan     1
Itaewon Clubs               1
Name: infection_case, dtype: int64

In [ ]:
# 전체 감염 경로
patientinfo['infection_case'].unique() # 29개

# 'overseas inflow', 'contact with patient', 'Seongdong-gu APT',
# 'etc', "Eunpyeong St. Mary's Hospital", 'Shincheonji Church',
# 'Dongan Church', 'Guro-gu Call Center', nan, 'Itaewon Clubs',
# 'Samsung Medical Center', 'Coupang Logistics Center',
# 'Guri Collective Infection', 'KB Life Insurance',
# 'Yeongdeungpo Learning Institute', 'Onchun Church',
# 'Cheongdo Daenam Hospital', 'Suyeong-gu Kindergarten',
# 'Ministry of Oceans and Fisheries', 'gym facility in Cheonan',
# 'gym facility in Sejong', 'River of Grace Community Church',
# 'Gyeongsan Seorin Nursing Home',
# 'Gyeongsan Cham Joeun Community Center',
# 'Gyeongsan Jeil Silver Town', 'Bonghwa Pureun Nursing Home',
# 'Pilgrimage to Israel', 'Milal Shelter', 'Geochang Church',
# 'Changnyeong Coin Karaoke'

 # 10. 연령대별 잉여 경로 비율

In [64]:
patientroute = pd.read_csv('./drive/My Drive/Colab Notebooks/Warmingup_Project/corona_visualization/PatientRoute.csv')
patientroute

patient_id  global_num  ...   latitude   longitude
0     1000000001         2.0  ...  37.615246  126.715632
1     1000000001         2.0  ...  37.567241  127.005659
2     1000000002         5.0  ...  37.592560  127.017048
3     1000000002         5.0  ...  37.591810  127.016822
4     1000000002         5.0  ...  37.563992  127.029534
...          ...         ...  ...        ...         ...
8087  6100000090         NaN  ...  37.558654  126.794474
8088  6100000090         NaN  ...  35.173220  128.946459
8089  6100000090         NaN  ...  35.336944  129.026389
8090  6100000090         NaN  ...  35.335757  129.025003
8091  6100000090         NaN  ...  35.336944  129.026389

[8092 rows x 8 columns]

In [ ]:
# 전체 이동 장소 타입
patientroute['type'].unique() # 25개

# 'airport', 'hospital', 'etc', 'store', 'public_transportation',
# 'restaurant', 'church', 'beauty_salon', 'pharmacy', 'pc_cafe',
# 'bank', 'academy', 'cafe', 'bakery', 'bar', 'gym', 'school',
# 'real_estate_agency', 'karaoke', 'post_office', 'gas_station',
# 'park', 'lodging', 'university', 'administrative_area_level_1'

In [67]:
essential = ['hospital', 'pharmacy', 'store', 'public_transportation', 'gas_station', 'administrative_area_level_1', 'governmnet_office', 'post_office', 
             'primary_school', 'school', 'university']
non_essential = ['library', 'theater', 'cafe', 'bakery', 'beauty_salon',
          'pc_cafe', 'billiards','restaurant', 'bar', 'karaoke', 'airport', 'lodging', 'academy', 'gym', 'mountain', 'park', 'police_office', 'bank', 'real_estate_agency',
          'church', 'etc']

In [131]:
def isessential(x):     
    if x in essential: 
        res = True
    else:
        res = False 
    return res

patientroute['non_essential'] = patientroute['type'].apply(lambda x: isessential(x))

a = patientroute[['patient_id', 'type', 'non_essential']]
a.head()

patient_id                   type  non_essential
0  1000000001                airport          False
1  1000000001               hospital           True
2  1000000002                    etc          False
3  1000000002                  store           True
4  1000000002  public_transportation           True

In [135]:
patientroute['route_value'] = 1
b = pd.pivot_table(data=patientroute, index = 'patient_id', columns = 'non_essential', values= 'route_value', aggfunc='sum', fill_value=0) # aggfunc = sum: 합계 구하기, fill_value=0 : 결측치 0 처리
b

non_essential  False  True 
patient_id                 
1000000001         1      1
1000000002         4      9
1000000003         2      0
1000000004         0      1
1000000005         0      1
...              ...    ...
6100000086         2      2
6100000087         2      6
6100000088        19      5
6100000089         1      1
6100000090         4      3

[1472 rows x 2 columns]

In [137]:
# Index로 있는 patient_id를 컬럼으로 빼내고 정리한다.
b.columns.name=None
b = b.reset_index()
b = b.rename(columns={False:'non_essential',True:'essential'})
b.head()

patient_id  non_essential  essential
0  1000000001              1          1
1  1000000002              4          9
2  1000000003              2          0
3  1000000004              0          1
4  1000000005              0          1

In [138]:
# 잉여 경로 비율 계산하기

b['surplus_ratio'] = b['non_essential'] / (b['non_essential'] + b['essential'])
b.head()

patient_id  non_essential  essential  surplus_ratio
0  1000000001              1          1       0.500000
1  1000000002              4          9       0.307692
2  1000000003              2          0       1.000000
3  1000000004              0          1       0.000000
4  1000000005              0          1       0.000000

In [139]:
# PatientInfo와 b(PatientRoute 요약)을 Inner Join 해서 가로 방향으로 붙여서 df_joined를 만든다.
df_joined = patientinfo.merge(b, left_on = 'patient_id', right_on = 'patient_id' , how='inner')
df_joined.head()

patient_id  global_num     sex  ... non_essential essential surplus_ratio
0  1000000001         2.0    male  ...             1         1      0.500000
1  1000000002         5.0    male  ...             4         9      0.307692
2  1000000003         6.0    male  ...             2         0      1.000000
3  1000000004         7.0    male  ...             0         1      0.000000
4  1000000005         9.0  female  ...             0         1      0.000000

[5 rows x 21 columns]

In [146]:
# 연령대별 잉여 경로 비율 구하기 (평균)

print('00대의 잉여 경로 비율은: ', df_joined[df_joined['age'] == '0s']['surplus_ratio'].mean() * 100,'%')
print('10대의 잉여 경로 비율은: ', df_joined[df_joined['age'] == '10s']['surplus_ratio'].mean() * 100,'%')
print('20대의 잉여 경로 비율은: ', df_joined[df_joined['age'] == '20s']['surplus_ratio'].mean() * 100,'%')
print('30대의 잉여 경로 비율은: ', df_joined[df_joined['age'] == '30s']['surplus_ratio'].mean() * 100,'%')
print('40대의 잉여 경로 비율은: ', df_joined[df_joined['age'] == '40s']['surplus_ratio'].mean() * 100,'%')
print('50대의 잉여 경로 비율은: ', df_joined[df_joined['age'] == '50s']['surplus_ratio'].mean() * 100,'%')
print('60대의 잉여 경로 비율은: ', df_joined[df_joined['age'] == '60s']['surplus_ratio'].mean() * 100,'%')
print('70대의 잉여 경로 비율은: ', df_joined[df_joined['age'] == '70s']['surplus_ratio'].mean() * 100,'%')
print('80대의 잉여 경로 비율은: ', df_joined[df_joined['age'] == '80s']['surplus_ratio'].mean() * 100,'%')

00대의 잉여 경로 비율은:  26.3664596273292 %
10대의 잉여 경로 비율은:  32.13268598685266 %
20대의 잉여 경로 비율은:  43.22467474149267 %
30대의 잉여 경로 비율은:  40.58549298252722 %
40대의 잉여 경로 비율은:  38.33934742819336 %
50대의 잉여 경로 비율은:  38.63739420015864 %
60대의 잉여 경로 비율은:  37.41564294019019 %
70대의 잉여 경로 비율은:  33.997150997150996 %
80대의 잉여 경로 비율은:  21.651315789473685 %


# ----------------------------------------------------------------------------------------------------------

# 결론

## 9. 연령대별 누적 확진자 수, 누적 사망자 수 및 사망률

### 00대 확진자 수 / 사망자 수 / 사망률 : 157 / 0 / 0%
### 10대 확진자 수 / 사망자 수 / 사망률 : 655 / 0 / 0%
### 20대 확진자 수 / 사망자 수 / 사망률 : 3176 / 0 / 0%
### 30대 확진자 수 / 사망자 수 / 사망률 : 1292 / 2 / 0.15%
### 40대 확진자 수 / 사망자 수 / 사망률 : 1521 / 3 / 0.19%
### 50대 확진자 수 / 사망자 수 / 사망률 : 2039 / 15 / 0.74%
### 60대 확진자 수 / 사망자 수 / 사망률 : 1405 / 39 / 2.78%
### 70대 확진자 수 / 사망자 수 / 사망률 : 725 / 80 / 11.03%
### 80대 확진자 수 / 사망자 수 / 사망률 : 498 / 131 / 26.31%

## 10. 연령대별 잉여 경로 비율 = 비필수경로/전체경로(어떤 나이대가 더 많이 쓸데없는 곳을 많이 가는 지) 보여주기

### 1위: 20대(43.23%)
### 2위: 30대(40.59%)
### 3위: 50대(38.64%)
### 4위: 40대(38.34%)
### 5위: 60대(37.42%)
### 6위: 70대(33.99%)
### 7위: 10대(32.13%)
### 8위: 00대(26.37%)
### 9위: 80대(21.65%)